### DISCUSSION TITLE
build docker images from nix

### DISCUSSION THREAD

**Discussion by brianmay** [teslamate-org/teslamate#4256] [Ideas]
@JakobLichterfeld 

Was curious how this would look. Not saying we should or shouldn't do this, just want to document what I did.

Used the treefmt pull request as a base.

```nix
        packages = {
          devenv-up = devShell.config.procfileScript;
          docker = pkgs.dockerTools.buildImage {
            name = "teslamate";
            tag = "latest";
            copyToRoot = pkg;
            config = {
              Cmd = ["${pkg}/bin/teslamate" "start"];
              ExposedPorts = {"4000/tcp" = {};};
            };
          };
          default = pkg;
        };
```

Obviously this is rough. I haven't looked at the entryPoint, the labels, or the grafana image. Maybe grafana should stay as is in fact, not sure.

Build with:

```console
$ nix build '.#docker'
```

This gives a tar.gz file:

```console
$ tar -tzvf result
dr-xr-xr-x root/root         0 1980-01-01 11:00 ./
-r--r--r-- root/root       534 1980-01-01 11:00 5cd122649c6ecb7f5dc2465620982e2ff3479d5ca6e78c4af944db5cf11e0df9.json
dr-xr-xr-x root/root         0 1980-01-01 11:00 e2e3bddfaeaf10145d0f8f2b0c417f6d8c4cd07e81533c884c43bcc7056d1743/
-r--r--r-- root/root         3 1980-01-01 11:00 e2e3bddfaeaf10145d0f8f2b0c417f6d8c4cd07e81533c884c43bcc7056d1743/VERSION
-r--r--r-- root/root       478 1980-01-01 11:00 e2e3bddfaeaf10145d0f8f2b0c417f6d8c4cd07e81533c884c43bcc7056d1743/json
-r--r--r-- root/root 574914560 1980-01-01 11:00 e2e3bddfaeaf10145d0f8f2b0c417f6d8c4cd07e81533c884c43bcc7056d1743/layer.tar
-r--r--r-- root/root       254 1980-01-01 11:00 manifest.json
-r--r--r-- root/root       102 1980-01-01 11:00 repositories
$ du -h /nix/store/cfhbkng9i4ln3kga47l1n84d3k1kgi6p-docker-image-teslamate.tar.gz
131M	/nix/store/cfhbkng9i4ln3kga47l1n84d3k1kgi6p-docker-image-teslamate.tar.gz
```

According to https://hub.docker.com/r/teslamate/teslamate/tags, the latest published image was 69.02MB + 55.25 MB + 66.79 MB. Or 191.24MB, the as expected the nix version is slightly smaller.

This image needs to be imported into docker. Use "docker load". See https://nix.dev/tutorials/nixos/building-and-running-docker-images.html for instructions. I haven't done that yet. Wonder if it actually is going to work.

**Comment by brianmay** [comment]
Not sure I understand the difference between setting `contents=` and `copyToRoot=`. But the build told me to use the later.

More examples here: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/docker/examples.nix

**Comment by JakobLichterfeld** [comment]
I love it. I have this on my list too. We've had issues in the past with the base image suddenly missing dependencies from time to time, and I really want to move the build to Nix.
